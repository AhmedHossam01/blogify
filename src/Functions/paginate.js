const numberPerPages = 6;

export const paginate = (posts, currentPage) => {
  const firstIndex = numberPerPages * currentPage;
  const lastIndex = firstIndex + numberPerPages;
  const currentPosts = posts.slice(firstIndex, lastIndex);
  return currentPosts;
};

export const getNumbersOfPages = posts => {
  const numberOfPages = Math.ceil(posts.length / numberPerPages);
  return numberOfPages;
};

export const getPages = posts => {
  const numberOfPages = getNumbersOfPages(posts);
  let pages = [];
  for (let i = 0; i < numberOfPages; i++) {
    pages.push(i);
  }
  return pages;
};
