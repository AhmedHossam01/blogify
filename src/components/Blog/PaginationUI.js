import React, { Component } from "react";
import Pagination from "react-bootstrap/Pagination";

export default class PaginationUI extends Component {
  constructor(props) {
    super(props);
  }

  prevPage = () => {
    if (this.props.currentPage > 0) {
      return this.props.changePage(this.props.currentPage - 1);
    }
  };

  nextPage = () => {
    if (this.props.currentPage < this.props.pages.length - 1) {
      return this.props.changePage(this.props.currentPage + 1);
    }
  };

  render() {
    const { changePage, currentPage, pages } = this.props;
    return (
      <div>
        <Pagination>
          <Pagination.First onClick={() => changePage(0)}></Pagination.First>
          <Pagination.Prev onClick={this.prevPage}></Pagination.Prev>
          {pages.map(number => {
            return (
              <Pagination.Item
                key={number}
                active={number === currentPage}
                onClick={() => changePage(number)}
              >
                {number}
              </Pagination.Item>
            );
          })}
          <Pagination.Next
            data-testid="next"
            onClick={this.nextPage}
          ></Pagination.Next>
          <Pagination.Last
            onClick={() => changePage(pages.length - 1)}
          ></Pagination.Last>
        </Pagination>
      </div>
    );
  }
}
