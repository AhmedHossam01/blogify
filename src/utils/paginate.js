const numberPerPage = 6;

export const paginate = (posts, currentPage) => {
  const firstIndex = numberPerPage * currentPage;
  const lastIndex = firstIndex + numberPerPage;
  const currentPosts = posts.slice(firstIndex, lastIndex);
  return currentPosts;
};

export const getNumbersOfPages = posts => {
  const numberOfPages = Math.ceil(posts.length / numberPerPage);
  return numberOfPages;
};

export const getPages = posts => {
  const numberOfPages = getNumbersOfPages(posts);
  let pages = [];
  for (let i = 0; i < numberOfPages; i++) {
    pages.push(i);
  }
  return pages;
};
