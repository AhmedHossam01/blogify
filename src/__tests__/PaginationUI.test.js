import React from "react";
import { mount } from "enzyme";
import PaginationUI from "../components/blog/PaginationUI";
import paginate from "../utils/paginate";

describe("PaginationUI use paginate function and renders correctly", () => {
  const fakePosts = [
    {
      userId: 1,
      id: 1,
      title:
        "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
      body:
        "quia et suscipitsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto"
    },
    {
      userId: 1,
      id: 2,
      title: "qui est esse",
      body:
        "est rerum tempore vitaesequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla"
    }
  ];

  jest.spyOn(paginate, "getPages");

  it("calls paginate function correctly", () => {
    const component = mount(
      <PaginationUI
        posts={fakePosts}
        numberPerPage={6}
        currentPage={0}
      ></PaginationUI>
    );

    // The paginate function is called one time
    expect(paginate.getPages.mock.calls.length).toBe(1);

    // The first argument of the first call of this function was fakePosts
    expect(paginate.getPages.mock.calls[0][0]).toBe(fakePosts);

    // It's clear!
    expect(paginate.getPages).toHaveBeenCalledWith(fakePosts, 6);

    expect(component).toMatchSnapshot();
    paginate.getPages.mockClear();
    component.unmount();
  });
});
