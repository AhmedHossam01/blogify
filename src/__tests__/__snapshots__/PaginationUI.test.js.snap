// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationUI use paginate function and renders correctly calls paginate function correctly 1`] = `
<PaginationUI
  currentPage={0}
  numberPerPage={6}
  posts={
    Array [
      Object {
        "body": "quia et suscipitsuscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto",
        "id": 1,
        "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
        "userId": 1,
      },
      Object {
        "body": "est rerum tempore vitaesequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla",
        "id": 2,
        "title": "qui est esse",
        "userId": 1,
      },
    ]
  }
>
  <div>
    <ForwardRef>
      <ul
        className="pagination"
      >
        <First
          onClick={[Function]}
        >
          <PageItem
            active={false}
            activeLabel="(current)"
            disabled={false}
            onClick={[Function]}
          >
            <li
              className="page-item"
            >
              <SafeAnchor
                className="page-link"
                disabled={false}
                onClick={[Function]}
              >
                <a
                  className="page-link"
                  href="#"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                >
                  <span
                    aria-hidden="true"
                  >
                    «
                  </span>
                  <span
                    className="sr-only"
                  >
                    First
                  </span>
                </a>
              </SafeAnchor>
            </li>
          </PageItem>
        </First>
        <Prev
          onClick={[Function]}
        >
          <PageItem
            active={false}
            activeLabel="(current)"
            disabled={false}
            onClick={[Function]}
          >
            <li
              className="page-item"
            >
              <SafeAnchor
                className="page-link"
                disabled={false}
                onClick={[Function]}
              >
                <a
                  className="page-link"
                  href="#"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                >
                  <span
                    aria-hidden="true"
                  >
                    ‹
                  </span>
                  <span
                    className="sr-only"
                  >
                    Previous
                  </span>
                </a>
              </SafeAnchor>
            </li>
          </PageItem>
        </Prev>
        <PageItem
          active={true}
          activeLabel="(current)"
          disabled={false}
          key="0"
          onClick={[Function]}
        >
          <li
            className="page-item active"
          >
            <span
              className="page-link"
              disabled={false}
              onClick={[Function]}
            >
              0
              <span
                className="sr-only"
              >
                (current)
              </span>
            </span>
          </li>
        </PageItem>
        <Next
          onClick={[Function]}
        >
          <PageItem
            active={false}
            activeLabel="(current)"
            disabled={false}
            onClick={[Function]}
          >
            <li
              className="page-item"
            >
              <SafeAnchor
                className="page-link"
                disabled={false}
                onClick={[Function]}
              >
                <a
                  className="page-link"
                  href="#"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                >
                  <span
                    aria-hidden="true"
                  >
                    ›
                  </span>
                  <span
                    className="sr-only"
                  >
                    Next
                  </span>
                </a>
              </SafeAnchor>
            </li>
          </PageItem>
        </Next>
        <Last
          onClick={[Function]}
        >
          <PageItem
            active={false}
            activeLabel="(current)"
            disabled={false}
            onClick={[Function]}
          >
            <li
              className="page-item"
            >
              <SafeAnchor
                className="page-link"
                disabled={false}
                onClick={[Function]}
              >
                <a
                  className="page-link"
                  href="#"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                >
                  <span
                    aria-hidden="true"
                  >
                    »
                  </span>
                  <span
                    className="sr-only"
                  >
                    Last
                  </span>
                </a>
              </SafeAnchor>
            </li>
          </PageItem>
        </Last>
      </ul>
    </ForwardRef>
  </div>
</PaginationUI>
`;
